// markdown-translatable/dist/components/block-translatable/BlockTranslatable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = require("@material-ui/core/styles");

var _core = require("@material-ui/core");

var _blockEditable = _interopRequireDefault(require("../block-editable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// const whyDidYouRender = (process.env.NODE_ENV !== 'production') ?
//   require('@welldone-software/why-did-you-render') : undefined;
// if (whyDidYouRender) whyDidYouRender(React);

/**
 * ### A reusable component for translating Markdown as HTML.
 * @component
 */
function BlockTranslatable(_ref) {
  var classes = _ref.classes,
      original = _ref.original,
      translation = _ref.translation,
      onTranslation = _ref.onTranslation,
      inputFilters = _ref.inputFilters,
      outputFilters = _ref.outputFilters,
      style = _ref.style,
      raw = _ref.raw;
  return _react.default.createElement(_core.Grid, {
    container: true,
    wrap: "nowrap",
    spacing: 0,
    className: classes.root,
    style: style
  }, _react.default.createElement(_core.Grid, {
    item: true,
    xs: 12,
    className: classes.original
  }, _react.default.createElement(_blockEditable.default, {
    markdown: original,
    inputFilters: inputFilters,
    outputFilters: outputFilters,
    raw: raw,
    editable: false
  })), _react.default.createElement(_core.Grid, {
    item: true,
    xs: 12,
    className: classes.translation
  }, _react.default.createElement(_blockEditable.default, {
    markdown: translation,
    inputFilters: inputFilters,
    outputFilters: outputFilters,
    onEdit: onTranslation,
    raw: raw
  })));
}

;
BlockTranslatable.propTypes = {
  /** Original markdown for the editor. */
  original: _propTypes.default.string.isRequired,

  /** Translation markdown for the editor. */
  translation: _propTypes.default.string.isRequired,

  /** Function to propogate changes to the translation. */
  onTranslation: _propTypes.default.func.isRequired,

  /** Replace strings before rendering. */
  inputFilters: _propTypes.default.array,

  /** Replace strings after editing. */
  outputFilters: _propTypes.default.array,

  /** CSS for the component. */
  style: _propTypes.default.object,

  /** Display Raw Markdown or HTML. */
  raw: _propTypes.default.bool
};
BlockTranslatable.defaultProps = {
  original: '',
  translation: '',
  onTranslation: function onTranslation() {},
  inputFilters: [],
  outputFilters: [],
  style: {},
  raw: false
};

var styles = function styles(theme) {
  return {
    root: {},
    original: {
      background: '#eee4'
    },
    translation: {}
  };
};

var areEqual = function areEqual(prevProps, nextProps) {
  var keys = ['original', 'translation', 'raw', 'style'];
  var checks = keys.map(function (key) {
    return JSON.stringify(prevProps[key]) === JSON.stringify(nextProps[key]);
  });
  var equal = !checks.includes(false); // console.log('BlockTranslatable', keys, checks, equal);

  return equal;
}; // BlockTranslatable.whyDidYouRender = true;


var StyleComponent = (0, _styles.withStyles)(styles)(BlockTranslatable);

var MemoComponent = _react.default.memo(StyleComponent, areEqual);

var _default = MemoComponent;
exports.default = _default;
