// markdown-translatable/dist/components/document-translatable/DocumentTranslatable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = require("@material-ui/core/styles");

var _md = _interopRequireDefault(require("md5"));

var _sectionTranslatable = _interopRequireDefault(require("../section-translatable"));

var helpers = _interopRequireWildcard(require("./helpers"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

// const whyDidYouRender = (process.env.NODE_ENV !== 'production') ?
//   require('@welldone-software/why-did-you-render') : undefined;
// if (whyDidYouRender) whyDidYouRender(React);

/**
 * ### A reusable component for translating a Markdown document.
 * @component
 */
function DocumentTranslatable(_ref) {
  var classes = _ref.classes,
      original = _ref.original,
      translation = _ref.translation,
      inputFilters = _ref.inputFilters,
      outputFilters = _ref.outputFilters,
      onTranslation = _ref.onTranslation,
      style = _ref.style;

  var _useState = (0, _react.useState)(0),
      _useState2 = _slicedToArray(_useState, 2),
      sectionFocus = _useState2[0],
      setSectionFocus = _useState2[1];

  var originalSections = helpers.sectionsFromMarkdown({
    markdown: original
  });

  var __translationSections = helpers.sectionsFromMarkdown({
    markdown: translation
  });

  var _useState3 = (0, _react.useState)(__translationSections),
      _useState4 = _slicedToArray(_useState3, 2),
      translationSections = _useState4[0],
      setTranslationSections = _useState4[1];

  var setTranslationSection = function setTranslationSection(_ref2) {
    var index = _ref2.index,
        translationSection = _ref2.translationSection;
    setSectionFocus(index);

    var _translationSections = _toConsumableArray(translationSections);

    _translationSections[index] = translationSection;
    setTranslationSections(_translationSections);

    var _translation = helpers.markdownFromSections({
      sections: _translationSections
    });

    onTranslation(_translation);
  };

  var sectionTranslatables = originalSections.map(function (originalSection, index) {
    var key = index + (0, _md.default)(JSON.stringify(originalSection));
    var translationSection = translationSections[index];

    var _onTranslation = function _onTranslation(_translationSection) {
      return setTranslationSection({
        index: index,
        translationSection: _translationSection
      });
    };

    var _onSectionFocus = function _onSectionFocus(expanded) {
      if (expanded) setSectionFocus(index);else setSectionFocus(null);
    };

    var _sectionFocus = sectionFocus === index;

    return _react.default.createElement(_sectionTranslatable.default, {
      key: key,
      original: originalSection,
      translation: translationSection,
      inputFilters: inputFilters,
      outputFilters: outputFilters,
      onTranslation: _onTranslation,
      onSectionFocus: _onSectionFocus,
      sectionFocus: _sectionFocus,
      style: style
    });
  });
  return _react.default.createElement(_react.default.Fragment, null, sectionTranslatables);
}

;
DocumentTranslatable.propTypes = {
  /** Original markdown for the editor. */
  original: _propTypes.default.string.isRequired,

  /** Translation markdown for the editor. */
  translation: _propTypes.default.string.isRequired,

  /** Function to propogate changes to the translation. */
  onTranslation: _propTypes.default.func.isRequired,

  /** Replace strings before rendering. */
  inputFilters: _propTypes.default.array,

  /** Replace strings after editing. */
  outputFilters: _propTypes.default.array,

  /** CSS for the component. */
  style: _propTypes.default.object
};
DocumentTranslatable.defaultProps = {
  original: '',
  translation: '',
  onTranslation: function onTranslation() {},
  inputFilters: [],
  outputFilters: [],
  style: {}
};

var styles = function styles(theme) {
  return {
    root: {},
    details: {
      display: 'block',
      padding: '0',
      borderTop: '1px solid #ccc',
      borderBottom: '1px solid #ccc'
    }
  };
};

var areEqual = function areEqual(prevProps, nextProps) {
  var keys = ['original', 'translation', 'style'];
  var checks = keys.map(function (key) {
    return JSON.stringify(prevProps[key]) === JSON.stringify(nextProps[key]);
  });
  var equal = !checks.includes(false); // console.log('DocumentTranslatable', keys, checks, equal);

  return equal;
}; // DocumentTranslatable.whyDidYouRender = true;


var StyleComponent = (0, _styles.withStyles)(styles)(DocumentTranslatable);

var MemoComponent = _react.default.memo(StyleComponent, areEqual);

var _default = MemoComponent;
exports.default = _default;
