// markdown-translatable/dist/components/block-editable/BlockEditable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styles = require("@material-ui/core/styles");

var helpers = _interopRequireWildcard(require("./helpers"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// const whyDidYouRender = (process.env.NODE_ENV !== 'production') ?
//   require('@welldone-software/why-did-you-render') : undefined;
// if (whyDidYouRender) whyDidYouRender(React);

/**
 * ### A reusable component for translating a Markdown block as HTML.
 * @component
 */
function BlockEditable(_ref) {
  var classes = _ref.classes,
      markdown = _ref.markdown,
      onEdit = _ref.onEdit,
      inputFilters = _ref.inputFilters,
      outputFilters = _ref.outputFilters,
      style = _ref.style,
      raw = _ref.raw,
      editable = _ref.editable;
  var component;

  var handleBlur = function handleBlur(_markdown) {
    var oldHTML = helpers.markdownToHtml({
      markdown: markdown,
      inputFilters: inputFilters
    });
    var newHTML = helpers.markdownToHtml({
      markdown: _markdown,
      inputFilters: inputFilters
    });
    if (oldHTML !== newHTML) onEdit(_markdown);
  };

  var handleHTMLBlur = function handleHTMLBlur(e) {
    var html = e.target.innerHTML;

    var _markdown = helpers.htmlToMarkdown({
      html: html,
      outputFilters: outputFilters
    });

    handleBlur(_markdown);
  };

  var handleRawBlur = function handleRawBlur(e) {
    var _markdown = helpers.filter({
      string: e.target.innerText,
      filters: outputFilters
    });

    handleBlur(_markdown);
  };

  if (raw) {
    var dangerouslySetInnerHTML = {
      __html: markdown
    };
    component = _react.default.createElement("div", {
      className: classes.markdown
    }, _react.default.createElement("pre", {
      className: classes.pre,
      dir: "auto",
      contentEditable: editable,
      onBlur: handleRawBlur,
      dangerouslySetInnerHTML: dangerouslySetInnerHTML
    }));
  } else {
    var _dangerouslySetInnerHTML = {
      __html: helpers.markdownToHtml({
        markdown: markdown,
        inputFilters: inputFilters
      })
    };
    component = _react.default.createElement("div", {
      className: classes.html,
      dir: "auto",
      contentEditable: editable,
      dangerouslySetInnerHTML: _dangerouslySetInnerHTML,
      onBlur: handleHTMLBlur
    });
  }

  return _react.default.createElement("div", {
    className: classes.root,
    style: style
  }, _react.default.createElement("div", {
    className: classes.wrapper
  }, component));
}

;
BlockEditable.propTypes = {
  /** @ignore */
  classes: _propTypes.default.object.isRequired,

  /** Initial markdown for the editor. */
  markdown: _propTypes.default.string.isRequired,

  /** Function to propogate changes to the markdown. */
  onEdit: _propTypes.default.func.isRequired,

  /** Replace strings before rendering. */
  inputFilters: _propTypes.default.array,

  /** Replace strings after editing. */
  outputFilters: _propTypes.default.array,

  /** CSS for the component. */
  style: _propTypes.default.object,

  /** Display Raw Markdown or HTML. */
  raw: _propTypes.default.bool,

  /** Enable/Disable editability. */
  editable: _propTypes.default.bool
};
BlockEditable.defaultProps = {
  markdown: '',
  onEdit: function onEdit() {},
  inputFilters: [],
  outputFilters: [],
  style: {},
  raw: false,
  editable: true
};

var styles = function styles(theme) {
  return {
    root: {
      height: '100%',
      width: '100%'
    },
    wrapper: {
      height: '100%',
      padding: '0 0.5em'
    },
    html: {
      height: '100%',
      width: '100%',
      display: 'grid'
    },
    markdown: {
      height: '100%',
      width: '100%',
      fontSize: '1.1em',
      display: 'grid'
    },
    pre: {
      whiteSpace: 'pre-wrap'
    }
  };
};

var areEqual = function areEqual(prevProps, nextProps) {
  var keys = ['markdown', 'raw', 'editable', 'style'];
  var checks = keys.map(function (key) {
    return JSON.stringify(prevProps[key]) === JSON.stringify(nextProps[key]);
  });
  var equal = !checks.includes(false); // console.log('BlockEditable', keys, checks, equal);

  return equal;
}; // BlockEditable.whyDidYouRender = true;


var StyleComponent = (0, _styles.withStyles)(styles)(BlockEditable);

var MemoComponent = _react.default.memo(StyleComponent, areEqual);

var _default = MemoComponent;
exports.default = _default;
