// markdown-translatable/dist/components/block-editable/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.filter = exports.htmlToMarkdown = exports.markdownToHtml = void 0;

var _showdown = _interopRequireDefault(require("showdown"));

var _turndown = _interopRequireDefault(require("turndown"));

var _turndownPluginGfm = require("turndown-plugin-gfm");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var turndownService = new _turndown.default({
  headingStyle: 'atx'
});
turndownService.use(_turndownPluginGfm.gfm);
turndownService.addRule('strikethrough', {
  filter: ['del', 's', 'strike'],
  replacement: function replacement(content) {
    return "~".concat(content, "~");
  }
});
turndownService.addRule('underline', {
  filter: ['u'],
  replacement: function replacement(content) {
    return "__".concat(content, "__");
  }
});
turndownService.addRule('emphasis', {
  filter: ['em'],
  replacement: function replacement(content) {
    return "*".concat(content, "*");
  }
});
var markdownToHtmlConverter = new _showdown.default.Converter();
markdownToHtmlConverter.setOption('underline', true);

var markdownToHtml = function markdownToHtml(_ref) {
  var markdown = _ref.markdown,
      _ref$inputFilters = _ref.inputFilters,
      inputFilters = _ref$inputFilters === void 0 ? [] : _ref$inputFilters;

  var _markdown = markdown.slice(0);

  _markdown = filter({
    string: _markdown,
    filters: inputFilters
  });
  var html = markdownToHtmlConverter.makeHtml(_markdown);
  return html;
};

exports.markdownToHtml = markdownToHtml;

var htmlToMarkdown = function htmlToMarkdown(_ref2) {
  var html = _ref2.html,
      _ref2$outputFilters = _ref2.outputFilters,
      outputFilters = _ref2$outputFilters === void 0 ? [] : _ref2$outputFilters;
  var markdown = turndownService.turndown(html);
  markdown = filter({
    string: markdown,
    filters: outputFilters
  });
  return markdown;
};

exports.htmlToMarkdown = htmlToMarkdown;

var filter = function filter(_ref3) {
  var string = _ref3.string,
      filters = _ref3.filters;

  var _string = string.slice(0);

  filters.forEach(function (filter) {
    var _filter = _slicedToArray(filter, 2),
        replacee = _filter[0],
        replacer = _filter[1];

    _string = _string.replace(replacee, replacer);
  });
  return _string;
};

exports.filter = filter;
